File: api/routers/library.py
"""Library endpoints for theological works."""

from fastapi import APIRouter, HTTPException, Query

from api.models.library import (
    Era,
    FiltersResponse,
    LibraryContextResponse,
    LibraryExpandMode,
    ReadingLevel,
    ScriptureRefsResponse,
    WorkDetail,
    WorkListResponse,
    WorkTOCResponse,
    WorkType,
)
from api.services import library_service

router = APIRouter(prefix="/library", tags=["library"])


# --- Works ---


@router.get("/works", response_model=WorkListResponse)
async def list_works(
    work_type: WorkType | None = Query(None, description="Filter by work type"),
    era: Era | None = Query(None, description="Filter by era"),
    reading_level: ReadingLevel | None = Query(None, description="Filter by reading level"),
    author: str | None = Query(None, description="Filter by author name"),
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
):
    """
    List all theological works in the library.

    Filter by work type, era, reading level, or author name.
    """
    return await library_service.get_works(
        work_type=work_type,
        era=era,
        reading_level=reading_level,
        author=author,
        limit=limit,
        offset=offset,
    )


@router.get("/works/{work_id}", response_model=WorkDetail)
async def get_work(work_id: str):
    """
    Get full details for a work.

    Includes author info, publication details, and content statistics.
    """
    work = await library_service.get_work(work_id)
    if not work:
        raise HTTPException(status_code=404, detail=f"Work not found: {work_id}")
    return work


@router.get("/works/{work_id}/toc", response_model=WorkTOCResponse)
async def get_work_toc(work_id: str):
    """
    Get table of contents for a work.

    Returns the full node tree structure WITHOUT leaf content.
    Use this for rendering navigation UI, sidebars, and book indexes.
    """
    toc = await library_service.get_work_toc(work_id)
    if not toc:
        raise HTTPException(status_code=404, detail=f"Work not found: {work_id}")
    return toc


# --- Nodes ---


@router.get("/works/{work_id}/nodes/{node_id}")
async def get_node(
    work_id: str,
    node_id: str,
    expand: LibraryExpandMode = LibraryExpandMode.NONE,
):
    """
    Get a specific node (chapter/section) from a work.

    Expand modes:
    - `none`: Node metadata only
    - `components`: Include resolved footnotes, images, quotes
    - `full`: Include scripture refs, author info, full navigation
    """
    node = await library_service.get_node(work_id, node_id, expand)
    if not node:
        raise HTTPException(
            status_code=404,
            detail=f"Node not found: {node_id} in work {work_id}",
        )
    # Return dict to avoid Union type serialization issues with Pydantic v2
    return node.model_dump()


# --- Scripture References ---


@router.get("/works/{work_id}/scripture-refs", response_model=ScriptureRefsResponse)
async def get_work_scripture_refs(
    work_id: str,
    book: str | None = Query(None, description="Filter by OSB book (e.g., 'matthew')"),
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
):
    """
    Get all scripture references from a work.

    Returns references to OSB passages with preview text.
    Filter by target book to find references to specific biblical books.
    """
    refs = await library_service.get_work_scripture_refs(work_id, book, limit, offset)
    if not refs:
        raise HTTPException(status_code=404, detail=f"Work not found: {work_id}")
    return refs


@router.get(
    "/works/{work_id}/nodes/{node_id}/scripture-refs",
    response_model=ScriptureRefsResponse,
)
async def get_node_scripture_refs(work_id: str, node_id: str):
    """
    Get scripture references from a specific node.

    Returns only references within the specified chapter/section.
    """
    refs = await library_service.get_node_scripture_refs(work_id, node_id)
    if not refs:
        raise HTTPException(
            status_code=404,
            detail=f"Node not found: {node_id} in work {work_id}",
        )
    return refs


# --- Filters ---


@router.get("/filters", response_model=FiltersResponse)
async def get_filters():
    """
    Get available filter values for the library index.

    Returns lists of authors, work types, eras, and reading levels
    that have at least one work in the library.
    """
    return await library_service.get_filters()


# --- Context (MCP) ---


@router.get("/context/{work_id}/{node_id}", response_model=LibraryContextResponse)
async def get_library_context(work_id: str, node_id: str):
    """
    Get rich context bundle for MCP/LLM consumption.

    Includes:
    - Node content with resolved components
    - Author information
    - Scripture references with OSB previews
    - Navigation links
    """
    context = await library_service.get_library_context(work_id, node_id)
    if not context:
        raise HTTPException(
            status_code=404,
            detail=f"Node not found: {node_id} in work {work_id}",
        )
    return context

File: reader_client/src/lib/api/library.ts
/**
 * Library API Client
 *
 * Handles communication with the library endpoints for theological works.
 */

import { api, type ApiRequestOptions } from './client';
import type {
	LibraryWorksResponse,
	LibraryWorkDetail,
	TocResponse,
	LibraryNode,
	LibraryNodeLeaf,
	LibraryExpandMode,
	LibraryScriptureRefsResponse,
	LibraryAuthorsResponse,
	LibraryAuthorDetail,
	LibraryAuthorWorksResponse,
	LibraryNodeContext,
	FiltersResponse,
	Era,
	WorkType,
	ReadingLevel
} from './types';

// ─────────────────────────────────────────────────────────────────────────────
// Works
// ─────────────────────────────────────────────────────────────────────────────

export interface ListWorksParams {
	work_type?: WorkType;
	era?: Era;
	reading_level?: ReadingLevel;
	author?: string;
	limit?: number;
	offset?: number;
}

export async function listWorks(
	params?: ListWorksParams,
	options?: ApiRequestOptions
): Promise<LibraryWorksResponse> {
	return api.get<LibraryWorksResponse>(
		'/library/works',
		params as Record<string, string | number | boolean | undefined>,
		options
	);
}

export async function getWork(
	workId: string,
	options?: ApiRequestOptions
): Promise<LibraryWorkDetail> {
	return api.get<LibraryWorkDetail>(`/library/works/${workId}`, undefined, options);
}

export async function getFilters(options?: ApiRequestOptions): Promise<FiltersResponse> {
	return api.get<FiltersResponse>('/library/filters', undefined, options);
}

export async function getToc(
	workId: string,
	options?: ApiRequestOptions
): Promise<TocResponse> {
	return api.get<TocResponse>(`/library/works/${workId}/toc`, undefined, options);
}

// ─────────────────────────────────────────────────────────────────────────────
// Nodes
// ─────────────────────────────────────────────────────────────────────────────

export async function getNode(
	workId: string,
	nodeId: string,
	expand?: LibraryExpandMode,
	options?: ApiRequestOptions
): Promise<LibraryNode> {
	return api.get<LibraryNode>(
		`/library/works/${workId}/nodes/${nodeId}`,
		expand ? { expand } : undefined,
		options
	);
}

/** Type-safe helper when you know you're fetching a leaf node with components */
export async function getLeafNode(
	workId: string,
	nodeId: string,
	options?: ApiRequestOptions
): Promise<LibraryNodeLeaf> {
	// Use 'full' expand mode to get navigation data along with components
	const node = await api.get<LibraryNode>(
		`/library/works/${workId}/nodes/${nodeId}`,
		{ expand: 'full' },
		options
	);
	if (!node.is_leaf) {
		throw new Error(`Expected leaf node but got container: ${nodeId}`);
	}
	return node as LibraryNodeLeaf;
}

// ─────────────────────────────────────────────────────────────────────────────
// Scripture References
// ─────────────────────────────────────────────────────────────────────────────

export interface ScriptureRefsParams {
	book?: string;
	limit?: number;
	offset?: number;
}

export async function getWorkScriptureRefs(
	workId: string,
	params?: ScriptureRefsParams,
	options?: ApiRequestOptions
): Promise<LibraryScriptureRefsResponse> {
	return api.get<LibraryScriptureRefsResponse>(
		`/library/works/${workId}/scripture-refs`,
		params as Record<string, string | number | boolean | undefined>,
		options
	);
}

export async function getNodeScriptureRefs(
	workId: string,
	nodeId: string,
	options?: ApiRequestOptions
): Promise<LibraryScriptureRefsResponse> {
	return api.get<LibraryScriptureRefsResponse>(
		`/library/works/${workId}/nodes/${nodeId}/scripture-refs`,
		undefined,
		options
	);
}

// ─────────────────────────────────────────────────────────────────────────────
// Authors
// ─────────────────────────────────────────────────────────────────────────────

export async function listAuthors(
	role?: string,
	options?: ApiRequestOptions
): Promise<LibraryAuthorsResponse> {
	return api.get<LibraryAuthorsResponse>(
		'/library/authors',
		role ? { role } : undefined,
		options
	);
}

export async function getAuthor(
	authorId: string,
	options?: ApiRequestOptions
): Promise<LibraryAuthorDetail> {
	return api.get<LibraryAuthorDetail>(`/library/authors/${authorId}`, undefined, options);
}

export async function getAuthorWorks(
	authorId: string,
	options?: ApiRequestOptions
): Promise<LibraryAuthorWorksResponse> {
	return api.get<LibraryAuthorWorksResponse>(
		`/library/authors/${authorId}/works`,
		undefined,
		options
	);
}

// ─────────────────────────────────────────────────────────────────────────────
// Context (MCP)
// ─────────────────────────────────────────────────────────────────────────────

export async function getNodeContext(
	workId: string,
	nodeId: string,
	options?: ApiRequestOptions
): Promise<LibraryNodeContext> {
	return api.get<LibraryNodeContext>(
		`/library/context/${workId}/${nodeId}`,
		undefined,
		options
	);
}

// ─────────────────────────────────────────────────────────────────────────────
// Namespace export for cleaner imports
// ─────────────────────────────────────────────────────────────────────────────

export const library = {
	listWorks,
	getWork,
	getFilters,
	getToc,
	getNode,
	getLeafNode,
	getWorkScriptureRefs,
	getNodeScriptureRefs,
	listAuthors,
	getAuthor,
	getAuthorWorks,
	getNodeContext
};
